### ISSUE TITLE
Server Side diff failing for fluent-bit

### ISSUE THREAD

**Issue by andrewjamesbrown** [argoproj/argo-cd#17568]
Checklist:

* [ ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

One of our ArgoCD instances is showing the following error when upgrading fluent-bit 0.30.4 -> 0.44.0:
```
Failed to compare desired state to live state: failed to calculate diff: error calculating server side diff: serverSideDiff error: error removing non config mutations for resource DaemonSet/fluent-bit: error reverting webhook removed fields in predicted live resource: .spec.template.spec.containers: element 0: associative list with keys has an element that omits key field "name" (and doesn't have default value)
```

**To Reproduce**
We use a Kyverno policy to modify images to point to a local ECR cache instead of pulling from dockerhub directly.
When upgrading the fluent-bit helm chart from 0.30.4 to 0.44.0, we get the error above.
We are using ArgoCD `v2.10.3+0fd6344`

**Expected behavior**

**Screenshots**

**Version**

```shell
v2.10.3+0fd6344
```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by bryanhorstmann** [comment]
Just ran into a similar issue with `kube-prometheus-stack`. I had to set `controller.diff.server.side: "false"` in order to unblock myself.  ArgoCD server: `v2.10.0+2175939`

I deleted a list from my values files and ran into this

**Comment by refl3ction** [comment]
I'm having the same issue.

I had to set `controller.diff.server.side: "true"` to fix another [error](https://github.com/argoproj/argo-cd/issues/17358) I was getting with `SyncOptions.ServerSideApply=true` on my applications, but now I'm getting this new error.

**Comment by algo7** [comment]
Same issue with 
```json
{
    "Version": "v2.10.5+335875d",
    "BuildDate": "2024-03-28T15:02:45Z",
    "GitCommit": "335875d13e018bed6e03873f4742582582964745",
    "GitTreeState": "clean",
    "GoVersion": "go1.21.3",
    "Compiler": "gc",
    "Platform": "linux/amd64",
    "KustomizeVersion": "v5.2.1 2023-10-19T20:13:51Z",
    "HelmVersion": "v3.14.3+gf03cc04",
    "KubectlVersion": "v0.26.11",
    "JsonnetVersion": "v0.20.0"
}

```

and `csi-driver-nfs-v4.6.0`

**Comment by ptr1120** [comment]
I am now having the same issue with our custom deployment and server-side diff activated:
```
Failed to compare desired state to live state: failed to calculate diff: error calculating server side diff: serverSideDiff error: error removing non config mutations for resource StatefulSet/xxx: error reverting webhook removed fields in predicted live resource: .spec.template.spec.containers: element 0: associative list with keys has an element that omits key field "name" (and doesn't have default value)
```
Even though my container 0 has a name. 

How to fix that?

**Comment by atlasoft** [comment]
Just ran into this and checking "apply only" allowed me to sync the application again.

![image](https://github.com/argoproj/argo-cd/assets/7682842/6974b9c8-311a-42d3-a533-1695cb55c203)

**Comment by Skaronator** [comment]
Just ran into this as well. We had server side apply and server-side diff enabled for a few weeks, and now it just broke.

The "apply only" mentioned above didn't help either.

Running the latest version, 2.11.0.

**Comment by Skaronator** [comment]
I found a workaround for my issue. I deleted the affected StatefulSet/Deployment with Orphan, which means it doesn't delete any pods. Then ran ArgoCD Sync again, which re-created the STS/Deploy resources.

**Comment by Skaronator** [comment]
Looks like my issue is more related to `ignoreDifferences` not working with `ServerSideDiff`. There is already an open issue: #17362

**Comment by STollenaar** [comment]
I think I found the issue related to ServerSide diff. Which is and issue inside the [gitops-engine](https://github.com/argoproj/gitops-engine) repo. Basically when you have a nested value being changed it breaks the map traversal used for doing these server side comparisons. I tried debugging it and applying a bandaid fix that I don't know if it would even be right https://github.com/argoproj/gitops-engine/commit/c25fd94b5dfcaf7d0a8020c2410a1d7629637b67#diff-00282c65a618a9ea64cdb99da5137663dc5773f2c3fd8c37ed2e9a99f3d67f09L254

**Comment by algo7** [comment]
Any update on this?

**Comment by gmauleon** [comment]
Having the same problem: `.spec.template.spec.containers: element 0: associative list with keys has an element that omits key field "name" (and doesn't have default value)`

Modified some containers environment variables in my spec.
ArgoCD 2.11.3 with server-side diff activated. Syncing manually fixed the problem.

But the main worry in our case is that's it's kind of a silent error, the app will flap from sync to unknown for a couple of minute and then back to `sync` for an hour every hour until we discovered it and manually `sync`. That last `bug` might be due to the fact that we use applicationset with progressive sync though.

**Comment by gmauleon** [comment]
Update: the error happens in a portion of the code that revert webhook mutation in the diffs.

So on our side adding the `IncludeMutationWebhook=true` in the already present compare-options annotation like so `argocd.argoproj.io/compare-options: ServerSideDiff=true,IncludeMutationWebhook=true`, bypass the error.

Not sure what other problems can arise down the line by setting this option though...

**Comment by adberger** [comment]
> Update: the error happens in a portion of the code that revert webhook mutation in the diffs as stated by @STollenaar
> 
> So on our side adding the `IncludeMutationWebhook=true` in the already present compare-options annotation like so `argocd.argoproj.io/compare-options: ServerSideDiff=true,IncludeMutationWebhook=true`, bypass the error.
> 
> Not sure what other problems can arise down the line by setting this option though...

Unknown Error state also gone?

**Comment by gmauleon** [comment]
Yes the unknown in this case was because of the errors while doing diffs, so there is definitely a problem in the code that "ignore the webhooks" in server side diffs, but so far it's a good workaround.

Including webhooks mutation in diffs will probably cause some unwanted differences though, depending on what webhooks you have in your clusters, in our case just ignoring the `/metadata/generation` at large did the trick.

```
      ignoreDifferences:
      - group: '*'
        jsonPointers:
        - /metadata/generation
        kind: '*'
````

**Comment by sstarcher** [comment]
I'm seeing similar issues for kube-prometheus-stack

```

ComparisonError: Failed to compare desired state to live state: failed to perform pre-diff normalization: error building typed results: error creating typedConfig: .spec.containers[1].ports: element 0: associative list with keys has an element that omits key field "protocol" (and doesn't have default value)
```

**Comment by andrii-korotkov-verkada** [comment]
ArgoCD versions 2.10 and below have reached EOL. Can you upgrade and tell us if the issue is still present, please?

**Comment by zswanson** [comment]
Observed as an issue in 2.13 too

**Comment by thecosmicfrog** [comment]
People experiencing this issue may want to follow my issue here, where @andrii-korotkov-verkada has been doing some digging into Kubernetes `structured-merge-diff`:

https://github.com/argoproj/argo-cd/issues/20792

I'm thinking we're all seeing roughly the same symptoms?

**Comment by cwstrommer** [comment]
@thecosmicfrog I'm inclined to concur that you're seeing the same issue from a different perspective.  Same goes for mine,  https://github.com/argoproj/argo-cd/issues/19994 , where the the block is renamed instead of removed - result being that  the block is duplicated.  I focused on the steps needed to consistently trigger this, but I agree that @andrii-korotkov-verkada has likely pinpointed the root cause.

**Comment by thecosmicfrog** [comment]
Hi all. A fix has been added to the `master` branch of Argo CD that has fixed the `[...] omits key field "name" (and doesn't have default value)` issue, at least for my reported issue:

https://github.com/argoproj/argo-cd/issues/20792#issuecomment-2537101349

I no longer get this error with server-side diff and server-side apply enabled. You may want to re-test.

Per @andrii-korotkov-verkada, you can build the latest image by cloning https://github.com/argoproj/argo-cd and running `make image`. I then re-tagged by running:

```bash
docker image tag argocd:latest <your-docker-repo>/argocd:issue-20792-fix
```

I then inserted this image into `argo-helm` by setting values as such:

```yaml
[...]

global:
  image:
    repository: <your-docker-repo>/argocd
    tag: issue-20792-fix

  # If uploaded to a private registry, you might need to add this:
  imagePullSecrets:
    - name: <secret-name>

[...]
```

...and `helm upgrade`'d my Argo CD.

Hopefully saves people a few minutes looking for the relevant values 🙂

**Comment by andrii-korotkov-verkada** [comment]
This has been fixed, I think latest 2.14 should already have the fix.
If not, would you be able to build an image from master and check if the error persists with server side diff on?
There's been a re-work of how server side diff works https://github.com/argoproj/gitops-engine/commit/4fd18478f5946b756fe253055358f89d4dd403a9, which should have made things better.

**Comment by alekhrycaiko** [comment]
I'm still seeing the behaviour on `2.14.4` where a nested value is changing and the server side diff is erroring out in kube-prometheus-stack. 

The error seems similar to what's posted in this issue, but let me know if this is sufficiently different and if a separate issue is warranted.
```
Failed to compare desired state to live state: failed to calculate diff: error calculating server side diff: serverSideDiff error: error removing non config mutations for resource PrometheusRule/kubernetes-monitoring-rules: error reverting webhook removed fields in predicted live resource: .spec.groups: element 0: associative list with keys may not have a null element
```

**Comment by andrii-korotkov-verkada** [comment]
Yeah, let's open a separate issue. Also, if you can, please try the build from master since it had some good rework of removing webhook mutations. Note that due to upcoming v3 release there would be some breaking changes.
